### ISSUE TITLE
Add Unit Testing for Services Business Logic

### ISSUE THREAD

**Issue by kspearrin** [bitwarden/server#419]
- [ ] Add unit testing of business logic in service layer
  - xUnit
  - NSubstitute
- [ ] Run tests on builds from AppVeyor CI

**Comment by xenji** [comment]
Hey @kspearrin! Can you provide a bit more context about how you want the tests to be structured? Should the tests for the respective service layers be in a eg. `Core.Tests` project? What is the coverage you aim for? What reports do you intend to generate from it? 
I'd might want to chime in and support (although being rather new to C# in general), but I might need a bit more guidance here.

**Comment by kspearrin** [comment]
@xenji Thanks for the interest. I likely need to do some initial project setup for this before a contributor can jump in to help. The end goal would be to provide as much coverage as possible, but it would probably make sense to prioritize things. I'll try to find some time soon to look at this a bit more and to get things set up with a few example tests that a contributor could follow. I'll likely also set up some coverage reporting as part of the CI build.

**Comment by JamesMcMahon** [comment]
Just a few thoughts and I hope these are useful as my goal to shared some of my experience and not to be pedantic.

I've been doing TDD for a little while and it is often difficult to layer unit tests onto existing code due to that code not being designed in a way to makes it easy to test. The general advice I've seen is to focus on integration tests and higher on the testing pyramid.

![image](https://user-images.githubusercontent.com/916676/50911072-b91acc80-13f4-11e9-98d5-8792032c845d.png)

(Note: this is just an example pyramid, it can vary from project to project)

Focusing on the testing layers above unit tests will usually give you the most bang for your buck initially. It will give you confidence that the existing features you have aren't going to regress as new code is added. It also gives you chance to codify and document the functionality of the existing system. 

As new features are created and existing code is refactored / modified. Then unit tests are added when needed. Micheal Feather's book Working Effectively With Legacy Code recommends the "sprout method",  which you can find described online at various places:

* http://taswar.zeytinsoft.com/learn-the-sprout-method-for-adding-new-functionality/
* https://www.codewithjason.com/taming-legacy-code-using-sprout-method-technique/

I think adding code coverage will give you a rough sense of if a PR is coming in with tests. Also it is a great first step to add the testing infrastructure to the code base like you just did. It will help encourage committers to write tests on new code. Continuously running those as part of CI will be an important signal to develop and will work in tandem with testing the existing functionality via higher level tests.

# TLDR:

* Use broad, higher level tests to document existing features and guard against regressions
* Adjust expectations and establish the needed infrastructure to encourage new code to be written with appropriate tests
* Add low level tests to existing code as it is refactored

That's my 2 cents. It would be great if I had some time to help by writing code but I will see what my schedule allows.

**Comment by kspearrin** [comment]
@JamesMcMahon The existing codebase is written with unit testing procedures in mind. There is a clear separation of concerns between the API interface, business logic (services), and data access (repositories). Dependency injection is used throughout to facilitate mocking/substitution. There may be some changes necessary to facilitate unit tests being written as it happens, but for the most part things should be pretty well conditioned for it.

I set up and created a pretty simple test here as a first example:
https://github.com/bitwarden/core/blob/master/test/Core.Test/Services/DeviceServiceTests.cs

I believe that what it really comes down to now is just figuring out what type of things and conditions we need to test for. The core server project here does not really have a ton of logic since most things are delegated to the client applications. For a lot of things, the server is really just an interface to the backing SQL database.
